File: ruby/grains/grains.rb
class Grains
  def self.square(number)
    raise ArgumentError if number < 1 || number > 64

    2 ** (number - 1)
  end

  def self.total
    (2 ** 64) - 1
  end
end

module BookKeeping
  VERSION = 1
end

File: javascript/grains/grains.js
const grainsOnPosition = max => {
  const result = [1];
  const positions = [...Array(max - 1).keys()];

  positions.forEach((_, index) => {
    result.push(result[BigInt(index)] * 2);
  });

  return result;
};

export const square = number => {
  if (number < 1 || number > 64) {
    throw new Error('square must be between 1 and 64');
  }

  return BigInt(grainsOnPosition(number)[number - 1]);
};

export const total = () => {
  return BigInt(
    grainsOnPosition(64).reduce((acc, cur) => BigInt(acc) + BigInt(cur))
  );
};
